src/storage/DataAccessorBasedStore.ts

```diff
@@ -0,0 +1,366 @@
+import type { Readable } from 'stream';
+import { DataFactory } from 'n3';
+import type { Quad } from 'rdf-js';
+import streamifyArray from 'streamify-array';
+import { v4 as uuid } from 'uuid';
+import type { Representation } from '../ldp/representation/Representation';
+import { RepresentationMetadata } from '../ldp/representation/RepresentationMetadata';
+import type { ResourceIdentifier } from '../ldp/representation/ResourceIdentifier';
+import { INTERNAL_QUADS } from '../util/ContentTypes';
+import { ConflictHttpError } from '../util/errors/ConflictHttpError';
+import { MethodNotAllowedHttpError } from '../util/errors/MethodNotAllowedHttpError';
+import { NotFoundHttpError } from '../util/errors/NotFoundHttpError';
+import { UnsupportedHttpError } from '../util/errors/UnsupportedHttpError';
+import type { MetadataController } from '../util/MetadataController';
+import { CONTENT_TYPE, HTTP, LDP, RDF } from '../util/UriConstants';
+import { ensureTrailingSlash, trimTrailingSlashes } from '../util/Util';
+import type { DataAccessor } from './accessors/DataAccessor';
+import type { ContainerManager } from './ContainerManager';
+import type { ResourceStore } from './ResourceStore';
+
+/**
+ * ResourceStore which uses a DataAccessor for backend access.
+ *
+ * The DataAccessor interface provides elementary store operations such as read and write.
+ * This DataAccessorBasedStore uses those elementary store operations
+ * to implement the more high-level ResourceStore contact, abstracting all common functionality
+ * such that new stores can be added by implementing the more simple DataAccessor contract.
+ * DataAccessorBasedStore thereby provides behaviours for reuse across different stores, such as:
+ *  * Converting container metadata to data
+ *  * Converting slug to URI
+ *  * Checking if addResource target is a container
+ *  * Checking if no containment triples are written to a container
+ *  * etc.
+ *
+ * Currently "metadata" is seen as something that is not directly accessible.
+ * That means that a user can't write directly to the metadata of a resource, only indirectly through headers.
```

Comment:
```suggestion
 * That means that a consumer can't write directly to the metadata of a resource, only indirectly through headers.
```

is that what we mean?